packages/versioning/src/lib.rs

```diff
@@ -1,115 +1,88 @@
-use serde::{Deserialize, Serialize};
-
 use sdk::{
-    cosmwasm_std::{to_binary, Binary, StdError, StdResult, Storage},
+    cosmwasm_std::{StdError, StdResult, Storage},
     cw_storage_plus::Item,
-    schemars::{self, JsonSchema},
 };
 
 pub type Version = u16;
 
-pub const VERSION_ITEM: Item<'static, Version> = Item::new("contract_version");
+pub type SemVer = (Version, Version, Version);
```

Comment:
I agree about the structure that it would be a bit lighter when both are changed, I am not sure whether that would be the case just for a software upgrade.

I strongly disagree with this approach because that should already be a function of the `migrate` function, not the version checking one. This leads me to the thought of renaming it to say something like `update_version_with_check` instead of `upgrade_contract` but again, I'd like to emphasize on the fact that putting a function call will not only break the single responsibility principle but also introduce a separate function call and call frame, when we already have a context, the `migrate` method, which is supposed to do that.